core/src/test/scala/org/apache/spark/scheduler/BlacklistTrackerSuite.scala

```diff
@@ -0,0 +1,503 @@
+    assertEquivalentToSet(tracker.isExecutorBlacklisted(_), Set())
+  }
+
+  test("blacklist can handle lost executors") {
+    // The blacklist should still work if an executor is killed completely.  We should still
+    // be able to blacklist the entire node.
+    val (tracker, scheduler) = trackerFixture
+    val taskSet0 = FakeTask.createTaskSet(4)
+    val tsm0 = new TaskSetManager(scheduler, Some(tracker), taskSet0, 4, clock)
+    // Lets say that executor 1 dies completely.  We get a task failure for the last task, but
+    // the taskset then finishes successfully (elsewhere).
+    (0 until 4).foreach { partition =>
+      tsm0.updateBlacklistForFailedTask("hostA", "1", partition)
+    }
+    tracker.handleRemovedExecutor("1")
+    tracker.updateBlacklistForSuccessfulTaskSet(0, 0, tsm0.execToFailures)
+    assert(tracker.isExecutorBlacklisted("1"))
+    clock.advance(tracker.BLACKLIST_TIMEOUT_MILLIS / 2)
+
+    // Now another executor gets spun up on that host, but it also dies.
+    val taskSet1 = FakeTask.createTaskSet(4, 1, 0)
+    val tsm1 = new TaskSetManager(scheduler, Some(tracker), taskSet1, 4, clock)
+    (0 until 4).foreach { partition =>
+      tsm1.updateBlacklistForFailedTask("hostA", "2", partition)
+    }
+    tracker.handleRemovedExecutor("2")
+    tracker.updateBlacklistForSuccessfulTaskSet(1, 0, tsm1.execToFailures)
+    // We've now had two bad executors on the hostA, so we should blacklist the entire node.
+    assert(tracker.isExecutorBlacklisted("1"))
+    assert(tracker.isExecutorBlacklisted("2"))
+    assert(tracker.isNodeBlacklisted("hostA"))
+
+    clock.advance(tracker.BLACKLIST_TIMEOUT_MILLIS / 2 + 1)
+    tracker.applyBlacklistTimeout()
+    // executor 1 is no longer explicitly blacklisted, since we've gone past its recovery time,
+    // but everything else is still blacklisted.
+    assert(!tracker.isExecutorBlacklisted("1"))
+    assert(tracker.isExecutorBlacklisted("2"))
+    assert(tracker.isNodeBlacklisted("hostA"))
+    // make sure we don't leak memory
+    assert(!tracker.executorIdToBlacklistStatus.contains("1"))
+    assert(!tracker.nodeToFailedExecs("hostA").contains("1"))
+    clock.advance(tracker.BLACKLIST_TIMEOUT_MILLIS)
+    tracker.applyBlacklistTimeout()
+    assert(!tracker.nodeIdToBlacklistExpiryTime.contains("hostA"))
+  }
+
+  test("task failures expire with time") {
+    val (tracker, scheduler) = trackerFixture
+    var stageId = 0
+    def failOneTaskInTaskSet(exec: String): Unit = {
+      val taskSet = FakeTask.createTaskSet(1, stageId, 0)
+      val tsm = new TaskSetManager(scheduler, Some(tracker), taskSet, 1, clock)
+      tsm.updateBlacklistForFailedTask("host-" + exec, exec, 0)
+      tracker.updateBlacklistForSuccessfulTaskSet(stageId, 0, tsm.execToFailures)
+      stageId += 1
+    }
+    failOneTaskInTaskSet("1")
+    // We have one sporadic failure on exec 2 -- it doesn't lead to an exec blacklist.
+    failOneTaskInTaskSet("2")
```

Comment:
why do you need the exec 2 failure here? it seems like the rest of the test deals with exec 1